packages/storage/amplify_storage_s3_dart/lib/src/exception/s3_storage_exception.dart

```diff
@@ -5,235 +5,82 @@
-  factory S3Exception.unknownException([String? extraMessage]) {
-    return S3Exception(
-      'Unknown exception occurred. $extraMessage',
-      recoverySuggestion: AmplifyExceptionMessages.missingExceptionMessage,
-    );
-  }
-
-  /// Creates a [S3Exception] that represents an unexpected error
-  /// when download operation doesn't receive all bytes when download completes.
-  factory S3Exception.incompleteDownload() {
-    return const S3Exception(
-      'Download has completed, but it has not received all bytes of content.',
-      recoverySuggestion: AmplifyExceptionMessages.missingExceptionMessage,
-    );
-  }
-
-  /// Creates a [S3Exception] that represents an unexpected error
-  /// when service API call doesn't return a valid content-length header.
-  factory S3Exception.unexpectedContentLengthFromService() {
-    return const S3Exception(
-      'Service has not returned a valid content length for requested object.',
-      recoverySuggestion: AmplifyExceptionMessages.missingExceptionMessage,
-    );
-  }
-
-  /// Creates a [S3Exception] that represents an expected exception
-  /// thrown when invoke cancel() on an eligible storage operation object.
-  factory S3Exception.controllableOperationCanceled() {
-    return const S3Exception(
-      'The requested operation has been canceled.',
-      recoverySuggestion:
-          'This is an expected exception when you call cancel() API on the storage operation object. You need to handle this exception to take further action on operation cancellation.',
-    );
-  }
-
-  /// An exception thrown when the `body` returned from GetObject is unexpectedly null.
-  factory S3Exception.unexpectedGetObjectBody() {
-    return const S3Exception(
-      'Unexpected null body from GetObject.',
-      recoverySuggestion: AmplifyExceptionMessages.missingExceptionMessage,
-    );
-  }
-
-  /// An exception created when attempt to resume a canceled operation.
-  factory S3Exception.resumeCanceledOperation() {
-    return const S3Exception(
-      'The operation has been canceled and can\'t be resumed.',
-      recoverySuggestion:
-          'You can only resume an operation that is paused by calling `pause()`.',
-    );
-  }
-
-  /// An exception thrown when the service doesn't return a valid uploadId on a
-  /// successful `CreateMultipartUploadRequest`.
-  factory S3Exception.unexpectedMultipartUploadId() {
-    return const S3Exception(
-      'Create multipart upload request succeeded, but it did not return a valid uploadId.',
-      recoverySuggestion: AmplifyExceptionMessages.missingExceptionMessage,
+/// The exception thrown when cancel a controllable operation.
+const s3ControllableOperationCanceledException = StorageOperationCanceled(
+  'The operation has been canceled.',
+  recoverySuggestion:
+      'This is an expected exception when you call cancel() API on the storage operation object. You need to handle this exception to take further action on operation cancellation.',
+);
```

Comment:
why return an exception if the cancel API was called and completed successfully?
in code we map `CancellationException` to `s3ControllableOperationCanceledException`. `CancellationException` is thrown when an operation is canceled before completion, which I think is due to some issue and not because of expected behavior of calling cancel API.

Reply:
Looking at the `operation.result` type `Future<Result>`, where `Result` is non-nullable.

If we don't throw exception, but complete the operation successfully on `cancel()` then `Result` has to be nullable, and return `null`, where developer would need a null-check even on a really successful operation.